transaction-context/src/transaction_accounts.rs

```diff
@@ -2,26 +2,249 @@
+    solana_account::{AccountSharedData, ReadableAccount, WritableAccount},
     solana_instruction::error::InstructionError,
     solana_pubkey::Pubkey,
     std::{
         cell::{Cell, UnsafeCell},
         ops::{Deref, DerefMut},
+        ptr,
+        sync::Arc,
     },
 };
 
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+struct SVMAccount {
+    key: Pubkey,
+    owner: Pubkey,
+    lamports: u64,
+    // The payload is going to be filled with the guest virtual address of the account payload
+    // vector.
+    // payload: VmSlice,
+}
+
+#[derive(Debug, PartialEq)]
+struct SVMAccountDeprecated {
+    rent_epoch: u64,
+    executable: bool,
+}
+
+struct PrivateAccountFields {
+    deprecated_fields: SVMAccountDeprecated,
+    payload: Arc<Vec<u8>>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct TransactionAccountView<'a> {
+    svm_account: &'a SVMAccount,
+    deprecated_fields: &'a SVMAccountDeprecated,
+    payload: &'a [u8],
+}
+
+impl ReadableAccount for TransactionAccountView<'_> {
+    fn lamports(&self) -> u64 {
+        self.svm_account.lamports
+    }
+
+    fn data(&self) -> &[u8] {
+        self.payload
+    }
+
+    fn owner(&self) -> &Pubkey {
+        &self.svm_account.owner
+    }
+
+    fn executable(&self) -> bool {
+        self.deprecated_fields.executable
+    }
+
+    fn rent_epoch(&self) -> u64 {
+        self.deprecated_fields.rent_epoch
+    }
+
+    fn to_account_shared_data(&self) -> AccountSharedData {
+        panic!(
+            "TransactionAccountView can't be transformed into AccountSharedData without losing \
+             reference constraints"
+        );
+    }
+}
+
+impl PartialEq<AccountSharedData> for TransactionAccountView<'_> {
+    fn eq(&self, other: &AccountSharedData) -> bool {
+        other.lamports() == self.lamports()
+            && other.data() == self.data()
+            && other.owner() == self.owner()
+            && other.executable() == self.executable()
+            && other.rent_epoch() == self.rent_epoch()
+    }
+}
+
+impl TransactionAccountView<'_> {
+    #[cfg(feature = "dev-context-only-utils")]
+    pub fn as_account_shared_data_for_tests(&self) -> AccountSharedData {
+        AccountSharedData::create(
+            self.lamports(),
+            self.data().to_vec(),
+            *self.owner(),
+            self.executable(),
+            self.rent_epoch(),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct TransactionAccountMutView<'a> {
+    svm_account: &'a mut SVMAccount,
+    deprecated_fields: &'a mut SVMAccountDeprecated,
+    payload: &'a mut Arc<Vec<u8>>,
+}
+
+impl TransactionAccountMutView<'_> {
+    fn data_mut(&mut self) -> &mut Vec<u8> {
+        Arc::make_mut(self.payload)
+    }
+
+    pub(crate) fn resize(&mut self, new_len: usize, value: u8) {
+        self.data_mut().resize(new_len, value)
+    }
+
+    #[cfg_attr(feature = "dev-context-only-utils", qualifiers(pub))]
+    pub(crate) fn set_data_from_slice(&mut self, new_data: &[u8]) {
```

Comment:
These functions that deal with the payload have been copied from the AccountSharedData crate, since we must perform the same job.